### DISCUSSION TITLE
Current landscape of branchless git projects

### DISCUSSION THREAD

**Discussion by hakanalpaydd** [arxanas/git-branchless#1401] [General]
I know this is a very open ended question, but as someone trying to recreate the easy VCS life I had while at Meta, I wanted to ask:

## What are the differences between the branchless git solutions (that you know about)?

- [git branchless](https://github.com/arxanas/git-branchless)
- [jujutsu](https://github.com/martinvonz/jj)
- [sapling](https://github.com/facebook/sapling)
- [stack pr](https://github.com/modularml/stack-pr)
- [ghstack](https://github.com/ezyang/ghstack)
- [git spr](https://github.com/ejoffe/spr)

non-FOSS lol:
- [graphite](https://graphite.dev)

I'm inclined to think that git branchless strikes the best balance of:
1. actually solving my problem: making my devx better
2. not being a huge overhead to learn
3. being fully compatible with my job's traditional github approach

However, git branchless doesn't have `hg split` which mercurial had at meta, and that made my life easier for sure. (which sapling has)

git branchless doesn't automatically name your branches something internal like sapling/mercurial and jujutsu do. This is nice for my unique situation for JIRA ticket integration. This is a difference I've noticed between this project and others

## And which would work best in my situation? 

I work at a company that has JIRA, so being able to name my branches `XXX-000_asablkjsd...` is nice for getting that automatic JIRA integration I guess. Also we use Aviator for our merge queue, so I'm not even sure if stack-pr's `land` would work since that assumes you can merge directly to master (?)

One thing I noticed that was nice from stack-pr is that they automatically add a PR message that displays the stack for reviewers to see:

![image](https://github.com/user-attachments/assets/86b84903-fc85-4b72-8188-f12b5a000481)

however I noticed all you really have to do to do this is:

---
1. #23655
2. -> #23537
---

with the PR numbers. And if you were to split a commit and resubmit the stack this message would be out of date anyways and need to be updated

**Comment by arxanas** [comment]
See https://github.com/arxanas/git-branchless/discussions/654 for a related discussion.

Tracking issue for `git split` is at https://github.com/arxanas/git-branchless/issues/180 if you want to try implementing it. You can also use both jj and git-branchless in the same repo to some degree, if you want to use its `jj split` command.

`git submit --forge github` will render a similar message and try to keep it up to date (but the GitHub integration is generally buggy; see https://github.com/arxanas/git-branchless/discussions/1259).
